Iter rework

an-owl opened the pull request:
pr for #19

- Defined `AddrIter` to panic when iterating over non-canonical addresses.
- Implemented `DoubleEndedIterator` and `ExactSizeIterator`
- Added a number of method implementations for iterator traits to reduce unnecessary calls to `Iterator::next()`
- Added an inclusive variant of the iterator. This is required because it cannot return the last address in a range without it. (e.g it will never return `0x7FFFFFFFFFFF` it will stop at `0x7FFFFFFFFFFE`).
- Added `From<Range>` and `From<RangeInclusive>` for `AddrIter`

8637ae6 requires that `MemoryAddress::RAW` implements `TryFrom<usize>`, I don't really like further restricting it but it already implements `TryInto<usize>` so it probably wont be an issue.

There may be bugs where iterating to a non-canonical address when `add`/`sub_checked()` returns `None`, by assuming its an overflow, when it should be checked. I think I caught them all but there might be more.

jounathaen commented on src/lib.rs:
```diff
@@ -144,21 +148,224 @@ impl<T: MemoryAddress> AddrRange<T> {
+trait IterInclusivity: 'static {
+    fn exhausted<T: Ord>(start: &T, end: &T) -> bool;
+}
+pub enum NonInclusive {}
```
Minor nitpicks: Could you please add a doc comment to these `pub` items? Also, the line spacing is a little inconsistent.

jounathaen commented on src/lib.rs:
```diff
@@ -144,21 +148,224 @@ impl<T: MemoryAddress> AddrRange<T> {
     }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        let Some(end) = self.end else {
```
Ah, nevermind. `end` is used later.

an-owl commented:
Sorry for the wait. I've added docs to the `pub` items with some doctests too which managed to catch a bug.

I also added some code comments where I thought they were necessary.

jounathaen commented:
Very nice! Once Clippy is satisfied, I'm as well :wink: 

I assume you are interested in a release after the merge?
